### PR TITLE
boards: b_u585i_iot02a: Add a non-secure target

### PR DESCRIPTION
**Pull request by erwango**
Provide a _ns target for b_u585i_iot02a board.

Tested using tfm_ipc sample.

### CODE REVIEW

**Comment by microbuilder** [comment]
I tried building this on OS X, but had to make the following modification to https://github.com/zephyrproject-rtos/trusted-firmware-m/blob/master/platform/ext/target/stm/common/scripts/postbuild.sh to point to the GNU version of `readlink` installed via `brew install coreutils`:

```bash
if [[ "$OSTYPE" == "darwin"* ]]; then
SCRIPT=$(greadlink -f $0)
else
SCRIPT=$(readlink -f $0)
fi
```

This then builds and runs the post build scripts, but when I run `west flash` it only seems to be flashing the NS binary:

```bash
$ west build -p auto -b b_u585i_iot02a_ns zephyr/samples/tfm_integration/tfm_ipc
...
[162/162] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       19224 B       512 KB      3.67%
            SRAM:       11304 B       786 KB      1.40%
        IDT_LIST:          0 GB         2 KB      0.00%
preprocess bl2 file
 updated
 updated
 updated
 updated
 updated
 updated
 updated
 updated
 updated
```

```bash
$ west flash
-- west flash: rebuilding
[0/14] Performing build step for 'tfm'
ninja: no work to do.
-- west flash: using runner stm32cubeprogrammer
-- runners.stm32cubeprogrammer: mass erase requested
      -------------------------------------------------------------------
                        STM32CubeProgrammer v2.9.0                  
      -------------------------------------------------------------------

ST-LINK SN  : 002A00363156501320323443
ST-LINK FW  : V3J7M3
Board       : B-U585I-IOT02A
Voltage     : 3.28V
SWD freq    : 24000 KHz
Connect mode: Under Reset
Reset mode  : Hardware reset
Device ID   : 0x482
Revision ID : Rev B
Device name : STM32U575/STM32U585
Flash size  : 2 MBytes
Device type : MCU
Device CPU  : Cortex-M33
BL Version  : 0x92
Debug in Low Power mode enabled


Mass erase ... 

Mass erase successfully achieved
      -------------------------------------------------------------------
                        STM32CubeProgrammer v2.9.0                  
      -------------------------------------------------------------------

ST-LINK SN  : 002A00363156501320323443
ST-LINK FW  : V3J7M3
Board       : B-U585I-IOT02A
Voltage     : 3.28V
SWD freq    : 24000 KHz
Connect mode: Under Reset
Reset mode  : Hardware reset
Device ID   : 0x482
Revision ID : Rev B
Device name : STM32U575/STM32U585
Flash size  : 2 MBytes
Device type : MCU
Device CPU  : Cortex-M33
BL Version  : 0x92
Debug in Low Power mode enabled



Memory Programming ...
Opening and parsing file: zephyr.hex
  File          : zephyr.hex
  Size          : 19224 Bytes
  Address       : 0x08074000 


Erasing memory corresponding to segment 0:
Erasing internal memory sectors [58 60]
Download in Progress:
[==================================================] 100% 

File download complete
Time elapsed during download operation: 00:00:00.227

RUNNING Program ... 
  Address:      : 0x8000000
Warning: The core is locked up
Start operation achieved successfully
```

This should be flashing `tfm_merged.hex`.

**Comment by erwango** [comment]
> This should be flashing ``

This is what I documented in board doc:
```
west flash --hex-file build/tfm_merged.hex
```

**Review by microbuilder — DISMISSED** [review]

**Comment by microbuilder** [comment]
> > This should be flashing `tfm_merged.hex`
> 
> This is what I documented in board doc:
> 
> ```
> west flash --hex-file build/tfm_merged.hex
> ```

Missed that, thanks. Looks good other than the CI failures, presumably due to the script hard-coding the toolchain, which should be `arm-zephyr-eabi-gcc` for the Zephyr SDK: https://github.com/zephyrproject-rtos/trusted-firmware-m/blob/master/platform/ext/target/stm/common/scripts/gcc/preprocess.sh#L23

**Comment by erwango** [comment]
> Missed that, thanks. Looks good other than the CI failures, presumably due to the script hard-coding the toolchain, which should be arm-zephyr-eabi-gcc for the Zephyr SDK: https://github.com/zephyrproject-rtos/trusted-firmware-m/blob/master/platform/ext/target/stm/common/scripts/gcc/preprocess.sh#L23

Ahhh, now I know what these CMake lines were supposed to do... Fixed.

File: boards/arm/b_u585i_iot02a/board.cmake [hunk 1]
```diff
@@ -1,4 +1,13 @@
 # SPDX-License-Identifier: Apache-2.0
+if(CONFIG_BUILD_WITH_TFM)
+  set(TFM_FLASH_BASE_ADDRESS 0x0C000000)
+
+  if (CONFIG_HAS_FLASH_LOAD_OFFSET)
```
Comment (microbuilder) [inline boards/arm/b_u585i_iot02a/board.cmake:5]: This should cause `west flash` to point to the correct .hex file when TF-M is enabled:

```suggestion
  # Flash merged TF-M + Zephyr binary
  set_property(TARGET runners_yaml_props_target PROPERTY hex_file "${CMAKE_BINARY_DIR}/tfm_merged.hex")

  if (CONFIG_HAS_FLASH_LOAD_OFFSET)
```

**Comment by microbuilder** [comment]
I noticed after flashing TF-M that you are unable to go back to a normal non TF-M application doing some variation of:

```bash
$ west build -p auto -b b_u585i_iot02a zephyr/samples/hello_world
$ west flash
```

The standard ST debug tool is unable to erase the flash memory from the secure region, and the previous TF-M image will continue to execute when you reset.

In order to 'recover' the TF-M device, you need to run 'build/tfm/regression.sh' once, and then you can rebuild the non-TFM image and flash it.

This should probably be added as a note to the board.rst file, or the correct command to erase the entire chip should be added if there is an alternative to running 'regression.sh', which is only available as a build artifact of a 'ns' application, and may not be obvious to new users testing out TF-M and going back to a non TF-M app.

File: boards/arm/b_u585i_iot02a/board.cmake (hunk 1 above)
Comment (erwango) [reply boards/arm/b_u585i_iot02a/board.cmake:5]: Working great, thanks!
Will update this board and provide the fix for others.

**Comment by erwango** [comment]
@microbuilder I've updated board configuration and documentation, taking into account your suggestions. Please let me know if this is fine and I'll then update other STM32 TFM targets.

**Review by microbuilder — APPROVED** [review]
Looks good, thanks.

File: boards/arm/b_u585i_iot02a/b_u585i_iot02a_ns.dts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			reg = <0x00074000 0x00080000>;
+		};
+		/*
+		 * The flash starting at 0xEA000 and ending at
+		 * 0xEB000 is reserved for the application.
+		 */
+		storage_partition: partition@EA000 {
+			label = "storage";
+			reg = <0x000EA000 0x00001000>;
+		};
```
Comment (FRASTM) [inline boards/arm/b_u585i_iot02a/b_u585i_iot02a_ns.dts:64]: with lower case ?

**Review by FRASTM — APPROVED** [review]
maybe the name of the PR should also refer to the stm32L5

**Comment by erwango** [comment]
Doc build error should likely be fixed with https://github.com/zephyrproject-rtos/zephyr/pull/41865

**Pull request merged**